File: java/src/test/java/bingo/BingoTest.java
package bingo;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.junit.jupiter.api.Test;

class BingoTest {

	BingoBoard board;

	@Test
	void aNewlyCreatedBoardIsNotInitialized() {
		givenBingoBoardOfSize(1, 1);
		thenBoardIsNotInitialized();
	}

	@Test
	void whenAllFieldsAreSetTheBoardIsInitialized() {
		String anyValue = "42";
		givenBingoBoardOfSize(1, 1);
		whenCellIsDefined(0, 0, anyValue);
		thenBoardIsInitialized();
	}

	@Test
	void whenAllFieldsOnRectangularBoardAreSetItIsInitialized() {
		String one = "one, two, three";
		String two = "Bingo cells can contain any text";
		givenBingoBoardOfSize(1, 2);
		whenCellIsDefined(0, 0, one);
		whenCellIsDefined(0, 1, two);
		thenBoardIsInitialized();
	}

	@Test
	void aDefinedCellCantBeRedefinedEvenIfItsTheSameValue() {
		String anyValue = "42";
		givenBingoBoardOfSize(1, 1);
		whenCellIsDefined(0, 0, anyValue);
		assertThatThrownBy(() -> whenCellIsDefined(0, 0, anyValue)) //
				.isInstanceOf(RuntimeException.class) //
				.hasMessageContaining("already defined");
	}

	@Test
	void duplicateCellsAreNotAllowed() {
		String anyValue = "42";
		givenBingoBoardOfSize(2, 2);
		whenCellIsDefined(0, 1, anyValue);
		assertThatThrownBy(() -> whenCellIsDefined(1, 0, anyValue)) //
				.isInstanceOf(RuntimeException.class) //
				.hasMessageContaining(anyValue + " already present at 0,1");
	}

	@Test
	void aNonInitializedBoardCannotBeMarked() {
		givenBingoBoardOfSize(1, 1);
		assertThatThrownBy(() -> whenCellIsMarked(0, 0)) //
				.isInstanceOf(RuntimeException.class) //
				.hasMessageContaining("not initialized");
	}

	@Test
	void whenAllCellGetsMarkedItIsMarked() {
		String anyValue = "42";
		givenBingoBoardOfSize(1, 1);
		whenCellIsDefined(0, 0, anyValue);
		whenCellIsMarked(0, 0);
		thenCellIsMarked(0, 0);
	}

	void givenBingoBoardOfSize(int width, int height) {
		board = new BingoBoard(width, height);
	}

	void whenCellIsDefined(int x, int y, String value) {
		board.defineCell(x, y, value);
	}

	void whenCellIsMarked(int x, int y) {
		board.markCell(x, y);
	}

	void thenBoardIsNotInitialized() {
		assertThat(boardInitializeState()).isFalse();
	}

	void thenBoardIsInitialized() {
		assertThat(boardInitializeState()).isTrue();
	}

	boolean boardInitializeState() {
		return board.isInitialized();
	}

	void thenCellIsMarked(int x, int y) {
		assertThat(board.isMarked(x, y)).isTrue();
	}

}

File: kotlin/src/test/kotlin/bingo/BingoTest.kt
package bingo

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test

/**
 * This class is identical to [BingoTest] but all BDD-style methods are
 * inlined (and thus this test does depend at multiple points on the
 * implementation details of [BingoBoard])
 */
internal class BingoTest {
    @Test
    fun `an newly created board is not initialized`() {
        val board = BingoBoard(1, 1)
        assertThat(board.isInitialized).isFalse()
    }

    @Test
    fun `when all fields are set the boar is initialized`() {
        val anyValue = "42"
        val board = BingoBoard(1, 1)
        board.defineCell(0, 0, anyValue)
        assertThat(board.isInitialized).isTrue()
    }

    @Test
    fun `when all fields on rectangular board are set it is initialized`() {
        val one = "one, two, three"
        val two = "Bingo cells can contain any text"
        val board = BingoBoard(1, 2)
        board.defineCell(0, 0, one)
        board.defineCell(0, 1, two)
        assertThat(board.isInitialized).isTrue()
    }

    @Test
    fun `a defined cell cant be redefined even if its the same value`() {
        val anyValue = "42"
        val board = BingoBoard(1, 1)
        board.defineCell(0, 0, anyValue)
        assertThatThrownBy { board.defineCell(0, 0, anyValue) } //
            .isInstanceOf(RuntimeException::class.java) //
            .hasMessageContaining("already defined")
    }

    @Test
    fun `duplicate cells are not allowed`() {
        val anyValue = "42"
        val board = BingoBoard(2, 2)
        board.defineCell(0, 1, anyValue)
        assertThatThrownBy { board.defineCell(1, 0, anyValue) } //
            .isInstanceOf(RuntimeException::class.java) //
            .hasMessageContaining("$anyValue already present at 0,1")
    }

    @Test
    fun `a non initialized board cannot be marked`() {
        val board = BingoBoard(1, 1)
        assertThatThrownBy { board.markCell(0, 0) } //
            .isInstanceOf(RuntimeException::class.java) //
            .hasMessageContaining("not initialized")
    }

    @Test
    fun `when all cell gets marked it is marked`() {
        val anyValue = "42"
        val board = BingoBoard(1, 1)
        board.defineCell(0, 0, anyValue)
        board.markCell(0, 0)
        assertThat(board.isMarked(0, 0)).isTrue()
    }
}
